cnxdb/migrations/20170911201035_transform_cnxml_to_html.py

```diff
@@ -0,0 +1,151 @@
+
+    batch_size = 10
+    logger.info('Batch size: {}'.format(batch_size))
+
+    start = time.time()
+    guesstimate = 1 * num_todo
+    guess_complete = guesstimate + start
+    logger.info('Completion guess: '
+                '"{}" ({})'.format(time.ctime(guess_complete),
+                                   timedelta(0, guesstimate)))
+
+    # Check if datamigrations.index_cnxml_html exists
+    cursor.execute("""\
+SELECT 1 FROM information_schema.tables
+    WHERE table_name = 'index_cnxml_html'
+      AND table_schema = 'datamigrations'""")
+    module_idents = tuple([i[1] for i in to_transform])
+    if not cursor.fetchall():
+        # Store module_files for modules we are going to update
+        cursor.execute("""\
+CREATE TABLE datamigrations.index_cnxml_html AS
+    SELECT * FROM module_files
+        WHERE module_ident IN %s
+          AND filename = 'index.cnxml.html'""",
+                       (module_idents,))
+    else:
+        cursor.execute("""\
+INSERT INTO datamigrations.index_cnxml_html
+    SELECT * FROM module_files
+        WHERE module_ident IN %s
+          AND filename = 'index.cnxml.html'
+          AND NOT EXISTS (
+            SELECT 1 FROM datamigrations.index_cnxml_html b
+                WHERE b.module_ident = module_files.module_ident);
+UPDATE datamigrations.index_cnxml_html b
+    SET fileid = module_files.fileid
+    FROM module_files
+    WHERE module_files.module_ident = b.module_ident
+      AND module_files.filename = b.filename
+      AND module_files.fileid != b.fileid
+      AND module_files.module_ident IN %s""",
+                       (module_idents, module_idents))
+
+    num_complete = 0
+    for batch in _batcher(to_transform, batch_size):
+        module_idents = tuple([i[1] for i in batch])
+        logger.info('Transform module_idents {}'.format(module_idents))
+
+        for fileid, module_ident in batch:
+            cursor.execute("""\
+WITH index_cnxml AS (
+    SELECT files.fileid, file
+        FROM module_files NATURAL JOIN files
+        WHERE module_ident = %(module_ident)s
+          AND filename = 'index.cnxml'
+        LIMIT 1
+), transformed AS (
+    SELECT html_content(%(module_ident)s) AS content
+) INSERT INTO files
+    (file, media_type)
+    SELECT convert_to(transformed.content, 'utf-8'), 'text/xml'
+        FROM index_cnxml, transformed
+        WHERE char_length(substring(encode(file, 'escape')
+                                    FROM 'cnxml-version=.0.7.')) > 0
+          AND NOT EXISTS (
+            SELECT 1 FROM files
+                WHERE sha1 = sha1(transformed.content))
+RETURNING fileid""", {'module_ident': module_ident})
+            new_fileid = cursor.fetchall()
+            if new_fileid:
+                cursor.execute("""\
+UPDATE module_files SET fileid = %s
+WHERE fileid = %s""", (new_fileid[0][0], fileid))
+
+        cursor.connection.commit()
+        num_complete += len(batch)
+        percent_comp = num_complete * 100.0 / num_todo
```

Comment:
Limiting precision on this makes for a more readable line: currently getting:
`[INFO] dbmigrator (20170911201035_transform_cnxml_to_html.py) - 43.689718859% complete est: "Thu Sep 14 09:33:43 2017" (0:25:24.662582)`
An easy way to limit it to 1 decimal is multiply by 1000, then divide the whole thing by 10.0